SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp

```diff
@@ -0,0 +1,412 @@
+
+
+
+namespace
+{
+	ULayeredLBStrategy* Strat;
+	UWorld* TestWorld;
+	TMap<FName, AActor*> TestActors;
+
+	// Copied from AutomationCommon::GetAnyGameWorld().
+	UWorld* GetAnyGameWorld()
+	{
+		UWorld* World = nullptr;
+		const TIndirectArray<FWorldContext>& WorldContexts = GEngine->GetWorldContexts();
+		for (const FWorldContext& Context : WorldContexts)
+		{
+			if ((Context.WorldType == EWorldType::PIE || Context.WorldType == EWorldType::Game)
+				&& (Context.World() != nullptr))
+			{
+				World = Context.World();
+				break;
+			}
+		}
+
+		return World;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND(FWaitForWorld);
+	bool FWaitForWorld::Update()
+	{
+		TestWorld = GetAnyGameWorld();
+
+		if (TestWorld && TestWorld->AreActorsInitialized())
+		{
+			AGameStateBase* GameState = TestWorld->GetGameState();
+			if (GameState && GameState->HasMatchStarted())
+			{
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	struct StrategyConfig {
+		TSubclassOf<UAbstractLBStrategy> StrategyClass;
+		TSet<TSoftClassPtr<AActor>> TargetActorTypes;
+	};
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FSetupStrategy,
+		std::vector<StrategyConfig>, ComponentStrategyConfigs, TOptional<uint32>, NumVirtualWorkers);
+	bool FSetupStrategy::Update()
+	{
+		ASpatialWorldSettings* WorldSettings = Cast<ASpatialWorldSettings>(TestWorld->GetWorldSettings());
+		WorldSettings->DefaultLayerLoadBalanceStrategy = UGridBasedLBStrategy::StaticClass();
+		WorldSettings->bEnableMultiWorker = true;
+
+		Strat = NewObject<ULayeredLBStrategy>(TestWorld);
+		int StratCounter = 0;
+		for (auto& StrategyConfig : ComponentStrategyConfigs)
+		{
+			auto strat_name = FName{ *FString(std::to_string(StratCounter++).c_str()) };
+			FLayerInfo layer_info;
+			layer_info.Name = strat_name;
+			layer_info.LoadBalanceStrategy = StrategyConfig.StrategyClass;
+
+			for (const auto& TargetActors : StrategyConfig.TargetActorTypes) {
+				layer_info.ActorClasses.Add(TargetActors);
+			}
+
+			WorldSettings->WorkerLayers.Add(strat_name, layer_info);
+		}
+
+		Strat->Init();
+
+
+		if (!NumVirtualWorkers) {
+			NumVirtualWorkers = Strat->GetMinimumRequiredWorkers();
+		}
+
+		Strat->SetVirtualWorkerIds(1, NumVirtualWorkers.GetValue());
+
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_ONE_PARAMETER(FSetupStrategysLocalWorker, VirtualWorkerId, worker_id);
+	bool FSetupStrategysLocalWorker::Update()
+	{
+		Strat->SetLocalVirtualWorkerId(worker_id);
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_THREE_PARAMETER(FCheckWhoShouldHaveAuthority, FAutomationTestBase*, Test, FName, ActorName, VirtualWorkerId, Expected);
+	bool FCheckWhoShouldHaveAuthority::Update()
+	{
+		VirtualWorkerId Actual = Strat->WhoShouldHaveAuthority(*TestActors[ActorName]);
+		Test->TestEqual(
+			FString::Printf(TEXT("Who Should Have Authority. Actual: %d, Expected: %d"), Actual, Expected),
+			Actual, Expected);
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND(FCleanup);
+	bool FCleanup::Update()
```

Comment:
Might be worth checking how some other tests do cleanup, specifically OwnershipLockingPolicyTest.cpp. It stores the test data as well as World* in a struct then has a deleter to do the cleanup instead of a cleanup function. Probably personal pref really since you have to create the struct in every test which kinda cancels out of the benefit of getting the remove all the Cleanup calls. On the other hand, you have to pass the struct data ref into every test, but you have less risk of data bleeding between tests I think.

Reply:
agreed that is a nice approach i'll look into it.

Reply:
done